### ISSUE TITLE
feat: Migrate codebase to Python 3.6+ (!70) - [merged]

### ISSUE THREAD

**Issue by kratsg** [histfitter/histfitter#67]
In GitLab by @feickert on Dec 17, 2020, 24:15

_Merges feat/migrate-to-Python-3.6-plus -> master_

Resolves #15 

~~This MR is still WIP for the time being as there are some ROOT v6.10 -> v6.20 transition pieces to work out to get the final `pytest` tests passing. These are understood though, and detailed below.~~

This MR should also get tested extensively, so I would like to request that people pull the Docker image

```
docker pull gitlab-registry.cern.ch/illinois/histfitter:latest
```

and have at it for stress testing and reproducing existing workflows. It would probably be a good idea to add **more** workflows to the pytest tests.

As this is going to be a **breaking change** I would recommend that a new release (~~`v0.66.0`~~`v0.67.0`) be made after this MR is merged.

cc the maintainers @jlorenz @llongo @pataraia for any feedback they have on this as it progresses.

```
* Migrate Python code to Python 3.6+
   - Use 2to3 and pyupgrade for automated sweeps
   - Remove vendored Python 2 argparse to prevent conflicts
   - Add requirements.txt files to define and constrain the runtime and development Python dependencies
   - Add optional pre-commit hooks for developers
   - Deprecate ROOT.Double for ctypes.c_double
* Update Dockerfile to support Python 3.6+ and ROOT v6.20
   - Use public AMG ATLAS Stats Docker images for base image
   - Use the recommended Kaniko executor workflow for building Docker images
   - Add linting workflow to CI using pre-commit and pyupgrade
* Update setup.sh and set minimum required ROOT version to v6.20
   - Apply linting suggestions from shellcheck
* Add test for intro of tutorial
```

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 24:24

Commented on [test/test_scripts.py line 19](https://github.com/histfitter/histfitter/compare/e9b030f047be63d0610c8bfc3b1f979518e20f9e..ea886d72c3386daf5f570ee88f842024ff0750c6#diff-3c33ca31de7662618d3d31acefc17cbbR19)



```
test/scripts/backupCache_input.py:131: FutureWarning: ROOT.Double is deprecated and will disappear in a future version of ROOT. Instead, use ctypes.c_double for pass-by-ref of doubles
hNorm.SetBinContent(1,h.IntegralAndError(0,h.GetNbinsX(),NormErr,""))
```

Is going to require the removal of `ROOT.Double` as a datatype and so is could potentially require a bit of extra testing.

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 24:24

Commented on [python/configManager.py line 1966](https://github.com/histfitter/histfitter/compare/1d1404af9346575169bcef253f6083d19debaf4f..289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#diff-1907f6f765a1075154a101c4b2efacfdR1966)



```
TypeError: '<' not supported between instances of 'TH1F' and 'TH1F'
```

Is another difference between ROOT versions. It isn't clear why the `sort` is needed though, so if this isn't actually **required** this can be safely ignored.

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 24:41

changed the description

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 24:52

Commented on [test/test_scripts.py line 19](https://github.com/histfitter/histfitter/compare/e9b030f047be63d0610c8bfc3b1f979518e20f9e..ea886d72c3386daf5f570ee88f842024ff0750c6#diff-3c33ca31de7662618d3d31acefc17cbbR19)



Also c.f. https://root-forum.cern.ch/t/pyroot-convert-ctypes-c-int-when-passing-it-as-argument/35805

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 01:32

added 1 commit

<ul><li>289f1d2a - Deprecate ROOT.Double() for ctypes.c_double()</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=714323&start_sha=62458a945aba43bf1579f6fd4b98977310283d37)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 01:33

changed the description

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 01:37

changed the description

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Dec 17, 2020, 01:40

Hi @feickert : many thanks for the proposed changes. As this is a so severe change I will take the time (probably tomorrow?) to run extended tests.

Jeanette

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 01:43

> As this is a so severe change I will take the time (probably tomorrow?) to run extended tests.

Fantastic and many thanks! I'm not expecting this to be a priority MR though as the holidays are so close and everyone is overburdened at the moment.

If there are extended tests that you can point me to that I can add please let me know and I will be happy to apply them.

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Dec 17, 2020, 01:46

what we typically do if we are doing something heavy, is that we run through the tutorial and make sure all the steps work there, potentially including the advanced tutorial (which has the benefit that one can update that one directly as well) - this is what I was planning to do. One problem that I see with a new tag at the moment, is that we have currently a non-understood issue with the histogram recycling, discussed in https://gitlab.cern.ch/HistFitter/HistFitter/-/merge_requests/67 that we should probably understand beforehand. 'cc @llongo here for the further discussion

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 09:52

Commented on [python/configManager.py line 1966](https://github.com/histfitter/histfitter/compare/1d1404af9346575169bcef253f6083d19debaf4f..289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#diff-1907f6f765a1075154a101c4b2efacfdR1966)



The [git blame on this also goes back 8 years to the directory creation](https://gitlab.cern.ch/HistFitter/HistFitter/-/blame/1d1404af9346575169bcef253f6083d19debaf4f/python/configManager.py#L1965), so there's no useful way to ask the author of the line

```python
histosToWrite.sort()
```

what they meant by a sorted list of `TH1F` objects (sorted by what?).

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 10:25

Commented on [python/configManager.py line 1966](https://github.com/histfitter/histfitter/compare/1d1404af9346575169bcef253f6083d19debaf4f..289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#diff-1907f6f765a1075154a101c4b2efacfdR1966)



Ah, after checking for myself I see that `sort` on `TH1F`s is sorting on their names alphabetically. This works in ROOT `v6.10`

```
$ docker run --rm -it gitlab-registry.cern.ch/histfitter/histfitter:v0.65.0 /bin/bash
builder@b7de7adbc7d8:/$ . /usr/local/bin/thisroot.sh
builder@b7de7adbc7d8:/$ root-config --version
6.10/07
builder@b7de7adbc7d8:/$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import ROOT
>>> hists = []
>>> hists.append(ROOT.TH1F("hist_b", "", 10, 0, 10))
>>> hists.append(ROOT.TH1F("hist_a", "", 10, 0, 10))
>>> print(hists)
[<ROOT.TH1F object ("hist_b") at 0x3c57960>, <ROOT.TH1F object ("hist_a") at 0x4003a20>]
>>> hists.sort()
>>> print(hists)
[<ROOT.TH1F object ("hist_a") at 0x4003a20>, <ROOT.TH1F object ("hist_b") at 0x3c57960>]
```

but is broken in ROOT `v6.20`

```
$ docker run --rm -it atlasamglab/stats-base:root6.20.06
root@3c54e5b40ad6:~/data# python
Python 3.8.3 (default, Jun  9 2020, 17:49:41) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ROOT
>>> hists = []
>>> hists.append(ROOT.TH1F("hist_b", "", 10, 0, 10))
>>> hists.append(ROOT.TH1F("hist_a", "", 10, 0, 10))
>>> print(hists)
[<ROOT.TH1F object ("hist_b") at 0x5557619dc490>, <ROOT.TH1F object ("hist_a") at 0x5557619deac0>]
>>> hists.sort()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'TH1F' and 'TH1F'
```

but fixed again in ROOT `v6.22`

```
$ docker run --rm -it atlasamglab/stats-base:root6.22.00
root@314decbb6b34:~/data# python
Python 3.8.3 (default, Jun  9 2020, 17:49:41) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ROOT
>>> hists = []
>>> hists.append(ROOT.TH1F("hist_b", "", 10, 0, 10))
>>> hists.append(ROOT.TH1F("hist_a", "", 10, 0, 10))
>>> print(hists)
[<cppyy.gbl.TH1F object at 0x5630d7cf4800>, <cppyy.gbl.TH1F object at 0x5630d7cf4410>]
>>> print([hist.GetName() for hist in hists])
['hist_b', 'hist_a']
>>> hists.sort()
>>> print([hist.GetName() for hist in hists])
['hist_a', 'hist_b']
```

and HistFitter currently segfaults in ROOT `v6.22`. :disappointed: 

I can try to build a release of `atlasamglab/stats-base` against the ROOT `v6.20.08` patchset and see if it is fixed there.

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 13:23

Commented on [python/configManager.py line 1966](https://github.com/histfitter/histfitter/compare/1d1404af9346575169bcef253f6083d19debaf4f..289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#diff-1907f6f765a1075154a101c4b2efacfdR1966)



Unfortunately the same thing happens on ROOT `v6.20.08`

```
root@38654cb60b22:~/data# root --version
ROOT Version: 6.20/08
Built for linuxx8664gcc on Dec 17 2020, 20:30:00
From tags/v6-20-08@v6-20-08
root@38654cb60b22:~/data# python
Python 3.8.6 (default, Dec 11 2020, 14:38:29) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ROOT
>>> letters = ["a", "b", "c", "d", "e"]
>>> hists = [ROOT.TH1F(f"hist_{letter}", "", 10, 0, 10) for letter in reversed(letters)]
>>> print(hists)
[<ROOT.TH1F object ("hist_e") at 0x55fded6a9f00>, <ROOT.TH1F object ("hist_d") at 0x55fdeb530f00>, <ROOT.TH1F object ("hist_c") at 0x55fdee8afbc0>, <ROOT.TH1F object ("hist_b") at 0x55fdee8a7fc0>, <ROOT.TH1F object ("hist_a") at 0x55fdee8bdf10>]
>>> hists.sort()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'TH1F' and 'TH1F'
```

but this can be solved easily across all version by just using `sorted` over `sort`

```
>>> import ROOT
>>> letters = ["a", "b", "c", "d", "e"]
>>> hists = [ROOT.TH1F(f"hist_{letter}", "", 10, 0, 10) for letter in reversed(letters)]
>>> sorted(hists, key=ROOT.TObject.GetName)
[<ROOT.TH1F object ("hist_a") at 0x55fdee8bdf10>, <ROOT.TH1F object ("hist_b") at 0x55fdee8a7fc0>, <ROOT.TH1F object ("hist_c") at 0x55fdee8afbc0>, <ROOT.TH1F object ("hist_d") at 0x55fdeb530f00>, <ROOT.TH1F object ("hist_e") at 0x55fded6a9f00>]
```

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 13:53

Commented on [python/configManager.py line 1966](https://github.com/histfitter/histfitter/compare/1d1404af9346575169bcef253f6083d19debaf4f..289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#diff-1907f6f765a1075154a101c4b2efacfdR1966)



changed this line in [version 3 of the diff](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715023&start_sha=289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0#680badf4c15b0f9a3d870be684c07854c562625c_1966_1962)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 13:53

added 1 commit

<ul><li>bb1699d1 - Use sorted with key option for broader ROOT support</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715023&start_sha=289f1d2afb2e74d73029ece73ddc38ce2e9aa0a0)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 14:00

resolved all threads

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 14:00

unmarked as a **Work In Progress**

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 14:04

All the ROOT `v6.10` -> ROOT `v6.20` migration issues that we showing up in the `pytest` tests are now resolved and so this MR is passing the same tests as before. We've already discussed that it should be run through the [![Tutorial](https://img.shields.io/badge/Tutorial-orange)](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/HistFitterTutorial), but it is now ready for review. :thumbsup:

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 14:04

changed the description

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 14:41

marked as a **Work In Progress**

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 15:27

added 1 commit

<ul><li>170d2b7d - Use sorted with key option for broader ROOT support</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715060&start_sha=bb1699d1b1058f0de3f3c996eb4e5755e3494742)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 15:27

unmarked as a **Work In Progress**

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 17:27

added 1 commit

<ul><li>ae12407a - Add test of analysis/tutorial/MyUserAnalysis.py</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715085&start_sha=170d2b7def7d8d466ee91ab04fc35f5e7281f637)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 17:45

changed the description

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 18:33

added 1 commit

<ul><li>d5c894c1 - Fix sorting of InputTrees with filename lambda</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715089&start_sha=ae12407af62b9e0b1abf82bf120d45977420d16a)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 21:19

added 4 commits

<ul><li>bde7baee - Deprecate ROOT.Double() for ctypes.c_double()</li><li>b023f2a0 - Use sorted with key option for broader ROOT support</li><li>6ce0618a - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>2ac367d6 - Fix sorting of InputTrees with filename lambda</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715112&start_sha=d5c894c18f96458e1a48f23be62f72170f5f4fa3)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 22:24

added 8 commits

<ul><li>b883b5ad - Reduce abuse of PYTHONPATH</li><li>ac65a21c - Set minimum compat ROOT in setup.sh</li><li>928382e4 - Deprecate ROOT.Double() for ctypes.c_double()</li><li>f39428f4 - Use sorted with key option for broader ROOT support</li><li>1d392cae - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>61788d1f - Fix sorting of InputTrees with filename lambda</li><li>4c3f0396 - Fix SysTable and UpperLimitTable namespace error</li><li>916b3815 - Fix writing of pickle binary</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715118&start_sha=2ac367d694479efa55de6a0720d2f8acfbe3d092)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 17, 2020, 22:42

@jlorenz @llongo After doing a quick run through the [![Tutorial](https://img.shields.io/badge/Tutorial-orange)](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/HistFitterTutorial) all the commands in it ran without error. I haven't taken the time to step through in more detail to investigate the output is identical in all cases, but things are looking pretty good. I've rebased this to pick up a few small fixes, so once you have done your own validation of the tutorial with this branch please let me know if you find anything that is outside of your expectations.

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 18, 2020, 24:02

added 4 commits

<ul><li>61afd8a5 - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>088d9ba8 - Fix sorting of InputTrees with filename lambda</li><li>b00f44e5 - Fix SysTable and UpperLimitTable namespace error</li><li>5c9eb7b0 - Fix writing of pickle binary</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715139&start_sha=916b3815f2cbf4fd54dc4ff37b5b7859bebe3461)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 18, 2020, 13:25

added 31 commits

<ul><li>5c9eb7b0...e9a3c383 - 13 commits from branch <code>HistFitter:master</code></li><li>d25ba029 - Add Python requirements.txt files</li><li>c42941da - Add pre-commit hooks</li><li>05b8c4ba - Apply 2to3 and pyupgrade to python/ dir</li><li>64c0bf28 - Apply 2to3 and pyupgrade to scripts/ dir</li><li>6df66b0a - Apply 2to3 and pyupgrade to test/ dir</li><li>fefadfd8 - Apply 2to3 and pyupgrade to analysis/ dir</li><li>858e38a3 - Apply 2to3 and pyupgrade to doxypy</li><li>ed8ee861 - Remove vendored Python 2.7 argparse</li><li>23f01fab - Migrate Dockerfile to Python 3.6+ and ROOT v6.20</li><li>ec13f2b4 - Update CI to use Kaniko workflow and add linting</li><li>9eb732e8 - Reduce abuse of PYTHONPATH</li><li>bb6f1c1b - Set minimum compat ROOT in setup.sh</li><li>a2a8684c - Deprecate ROOT.Double() for ctypes.c_double()</li><li>478f051b - Use sorted with key option for broader ROOT support</li><li>116db070 - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>50b10966 - Fix sorting of InputTrees with filename lambda</li><li>d1d65d84 - Fix SysTable and UpperLimitTable namespace error</li><li>467ba6dc - Fix writing of pickle binary</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715820&start_sha=5c9eb7b06027c8c8f14ce0a1059a5200d13c497f)

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 18, 2020, 13:27

This has now been rebased against `master` after the merge of MR !69.

**Comment by kratsg** [comment]
In GitLab by @feickert on Dec 18, 2020, 14:01

added 9 commits

<ul><li>8cd4a169 - Update CI to use Kaniko workflow and add linting</li><li>a4bddd95 - Reduce abuse of PYTHONPATH</li><li>abfb9eb4 - Set minimum compat ROOT in setup.sh</li><li>56f1f1e6 - Deprecate ROOT.Double() for ctypes.c_double()</li><li>d300eb21 - Use sorted with key option for broader ROOT support</li><li>978c65d6 - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>544e7d36 - Fix sorting of InputTrees with filename lambda</li><li>eb42e74e - Fix SysTable and UpperLimitTable namespace error</li><li>489d602f - Fix writing of pickle binary</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=715835&start_sha=467ba6dc478cb7f591607a4b3c17f9357f0cfb21)

**Comment by kratsg** [comment]
In GitLab by @feickert on Jan 7, 2021, 06:47

@jlorenz @llongo @pataraia Happy New Year! I wanted to follow up on this MR to see if there are additional tests that I can perform to share, or if you HistFitter experts can perform checks on known pathological workspaces for regressions.

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Jan 7, 2021, 07:02

Hi @feickert ,

happy new year!

No, from your side everything is fine - I just wanted to test it myself and thought to get to it before Christmas, but decided then that I should first focus on writing those applications and this exam...

I estimate to get to it during middle of next week.

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @feickert on Jan 7, 2021, 07:26

> but decided then that I should first focus on writing those applications and this exam

A wise choice. :)

> I estimate to get to it during middle of next week.

Fantastic! If anything is unclear please do not hesitate to ping me for an explanation, as I'm happy to try to step through the motivations for any changes.

**Comment by kratsg** [comment]
In GitLab by @gstark on Jan 26, 2021, 16:32

Hi @jlorenz , I know you're super busy :) Did you manage to check this out on your end locally to see if everything is ok? I would be very eager to move towards python3/various migrations.

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Feb 1, 2021, 07:56

Hi @gstark,

now I'm at testing on lxplus (locally I have unfortunately some Root version with python 3 mismatch). So far, the tutorial does not run smoothly for me - the step

HistFitter.py -t -w -f -F excl -D "before,after" -i analysis/tutorial/MyOneBinExample.py 

fails, with the error message 

Traceback (most recent call last):
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/scripts/HistFitter.py", line 355, in <module>
    configMgr.executeAll()
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/configManager.py", line 947, in executeAll
    self.execute(tl)
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/configManager.py", line 1012, in execute
    self.setWeightsCutsVariable(chan, sam, regionString)
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/configManager.py", line 1471, in setWeightsCutsVariable
    self.prepare.read(sam.input_files, suffix=sam.getTreenameSuffix(), friendTreeName=sam.friendTreeName)
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/prepareHistos.py", line 273, in read
    self.configMgr.chains[chainID].Project._creates = True
AttributeError: 'cppyy.CPPOverload' object has no attribute '_creates'
Exception ignored in: <function PrepareHistos.__del__ at 0x7f30f00b1b90>
Traceback (most recent call last):
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/prepareHistos.py", line 121, in __del__
AttributeError: 'PrepareHistos' object has no attribute 'cacheFile'

Strange, that @feickert didn't see this during testing. I'm using Root 6.22/00 setup with lsetup "views LCG_98python3 x86_64-centos7-gcc9-opt"

To be investigated what is wrong....

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Feb 1, 2021, 10:44

Hi @feickert 

to give an update on this, and after discussing with @gstark : the problem seems to arise from that for certain python 2.6 features being dropped with python 3. This relates to the _creates property that was available formerly, but isn't there now anymore. (We suspect that the support was dropped because of an improved memory handling.) Removing anything in direction of _creates in the code in python/prepareHistos.py ensures that the (basic) tutorial runs fine, and I obtain good results. (I am now testing the advanced tutorial.) However, given that this code part is particularly delicate, I suggest that @llongo also explicitly tests the histogram recycling functionality.

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Feb 1, 2021, 10:54

Hi,

and finally, also the advanced tutorial runs fine.

However, I still have one error message left if using histogram building:

Exception ignored in: <function PrepareHistos.__del__ at 0x7fb3700bab90>
Traceback (most recent call last):
  File "/afs/cern.ch/user/j/jlorenz/HistFitter_testpython3/HistFitter/python/prepareHistos.py", line 121, in __del__
AttributeError: 'PrepareHistos' object has no attribute 'cacheFile'

This could just hint to that the cache file is deleted before getting to this delete function, but @llongo's check should be able to give here a bit more insight.

From my side, given the mentioned modifications, this looks fine. @feickert thanks again for preparing!

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @feickert on Feb 2, 2021, 24:14

Thanks very much for taking the time to perform these tests and step through to give this feedback, @jlorenz &mdash; I truly appreciate it. I'll note that given some issues I've had with ROOT 6.22 (which seems to create trouble across multiple libraries) I've only used ROOT `v6.20` in my testing so far.

```
$ docker pull gitlab-registry.cern.ch/illinois/histfitter:latest
$ docker run --rm gitlab-registry.cern.ch/illinois/histfitter:latest "root --version && echo && python --version --version"
ROOT Version: 6.20/06
Built for linuxx8664gcc on Jun 26 2020, 04:23:00
From tags/v6-20-06@v6-20-06

Python 3.8.3 (default, Jun  9 2020, 17:49:41) 
[GCC 8.3.0]
```

@llongo if there are tests that I can help perform or look at with you for the histogram building that @jlorenz brings up please let me know. @jlorenz, can you also let us know where in particular in the [HistFitter Advanced Tutorial](https://twiki.cern.ch/twiki/bin/view/AtlasProtected/HistFitterAdvancedTutorial) you're seeing this, so I can start to look at replication tests/studies of the problem?

**Comment by kratsg** [comment]
In GitLab by @feickert on Feb 16, 2021, 23:07

Hi @llongo, I just wanted to ping on this to check if you've had any time to review it. If it would help to have some additional validation examples beyond the tutorials I can see if @gstark has some workspaces that I can validate against. Please let me know if you have any questions.

**Comment by kratsg** [comment]
In GitLab by @llongo on Feb 17, 2021, 02:07

Hi @feickert,
sorry for the no reaction. I'll try to do it by the end of this week.
Cheers,
Luigi

**Comment by kratsg** [comment]
In GitLab by @llongo on Feb 23, 2021, 12:05

Hi @feickert @jlorenz,
so I've finally done the test. The test has been done on lxplus using `lsetup "root 6.22.00-python3-x86_64-centos7-gcc8-opt"` in order to run it I need to slightly change these lines of code
```
self.configMgr.chains[chainID].Project._creates = True
self.configMgr.chains[chainID].Draw._creates = True
```
to 
```
ROOT.SetOwnership(self.configMgr.chains[chainID], True)
```
I've got the same fit results of running with the HF master and with `lsetup "root 6.18.04-x86_64-centos7-gcc8-opt"`. It was a very big job, thanks a lot Matthew.
The only thing is a bit worrying me is that we can not run without python3. On lxplus I was able to run only with the root version I mention at the beginning. From what I see this seems to come from the use of f-string in python3. While using format this would have not broken the python2 compatibility (or at least this is what I've understood).
Given the amount of work done, I would create a new tag with with the commits we have up to now in the master and then we could merge this and create a new tag for python3 (including the fixes above). @jlorenz what do you think?
Cheers, 
Luigi

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Feb 23, 2021, 12:09

Hi @llongo 

I agree with the proposed procedure, and then the next step will be to move the python 3-based version to github. I believe before doing so, we need to a minimal clean-up regarding ATLAS-internal stuff, possible move contour plotting related scripts to an extra package, and then update the copyright issues (tagging here @gstark with who I started to discuss about this). This will then be the next ticket/issue to work on.

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @gstark on Feb 23, 2021, 12:45

No issue here with migrating it over :) Let's see if we can't do the tags up to python 3 + fix licensing in gitlab first, and then migrate.

**Comment by kratsg** [comment]
In GitLab by @feickert on Feb 23, 2021, 14:36

added 1 commit

<ul><li>3d3a34ac - Give ROOT ownership of configMgr.chains[chainID]</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=753960&start_sha=489d602f4a1d5a5e1a55dbf644ca0848bece44e5)

**Comment by kratsg** [comment]
In GitLab by @feickert on Feb 23, 2021, 15:00

Thanks for the review @llongo. :thumbsup: If the only changes you needed were the replacement of those two lines then 3d3a34a should address that. Please let me know if there is additional fixes 

> The only thing is a bit worrying me is that we can not run without python3. On lxplus I was able to run only with the root version I mention at the beginning. From what I see this seems to come from the use of f-string in python3. While using format this would have not broken the python2 compatibility (or at least this is what I've understood)

Sure. This is something that is up to the maintainers to decide on, but from my experience with doing the Python 2 and Python 3 support dance for other libraries it just leads to extra work and pain for everyone (including users when they are forced to quickly switch when things really really break). I think @gstark will agree with me here when I say that just making a clean break and going to modern Python syntax (Python 3.6+) will simplify things for users and developers. Python 2 is EOL since 2020-01-01 and as all of ATLAS won't transition off it at the same time the more libraries and tools that can drop it and move forward the better for everyone. :thumbsup: 

To be clear, this isn't me trying to say that you're arguing to keep Python 2 support forever or anything silly like that. I'm just explaining from my views and experiences why attempting to keep Python 2 support isn't a good thing for a project.

All that being said, if it is important to keep Python 2 compatibility let me know and I can go through and migrate things back on another branch and then force-push it here.

> I would create a new tag with with the commits we have up to now in the master and then we could merge this and create a new tag for python3 (including the fixes above)

This seems like a good idea to me. :thumbsup: Definitely nice to have a tagged point before a breaking change. If there are other open MRs at the moment that you would want in before this breaking change and are all ready to go I have no problems with rebasing this MR on top &mdash; I rebased it many many times when I was working across multiple branches to do testing and cleaning.

> and then the next step will be to move the python 3-based version to github.

Hey! That's so cool to hear that HistFitter will be moving to GitHub (and presumably be open)! Congrats to the maintainers on this move. :rocket: :tada: I'm very much looking forward to seeing this, and please let me know if I can help with any of this migration.

> Let's see if we can't do the tags up to python 3 + fix licensing in gitlab first, and then migrate.

This would be great, IMO. :slight_smile:

**Comment by kratsg** [comment]
In GitLab by @feickert on Mar 10, 2021, 12:50

:wave: Hi @llongo @jlorenz. I just wanted to follow up on your plans/thoughts here:

> I would create a new tag with with the commits we have up to now in the master and then we could merge this and create a new tag for python3 (including the fixes above)

> I agree with the proposed procedure, and then the next step will be to move the python 3-based version to github.

and see if there is anything I can do to help with this (or the work that would follow after). :thumbsup:

**Comment by kratsg** [comment]
In GitLab by @jlorenz on Mar 11, 2021, 01:50

Hi @feickert 

the tag is on the side of @llongo , so I let him comment.
For the moving to github, @gstark and I found that there is somehow a private mirror of HistFitter on github, but this story is a bit mysterious, as the owner is not visible. We are in contact with the github team to sort this out and to then start the transition.

Cheers,
Jeanette

**Comment by kratsg** [comment]
In GitLab by @llongo on Mar 11, 2021, 11:54

Hi @feickert , 
I need just some extra days to include a couple of other things reported in the issue. Once it is done, I'll create the tag and ask you to update few things (few lines don't worry ;-) ).
Cheers, 
Luigi

**Comment by kratsg** [comment]
In GitLab by @feickert on Mar 11, 2021, 12:56

> to include a couple of other things reported in the issue

@llongo sorry, which issue? I think everything discussed in Issue #15 is resolved here.

> ask you to update few things (few lines don't worry ;-)

Sure sounds good. If there are things that I can start to address now I'm happy to do so.

**Comment by kratsg** [comment]
In GitLab by @llongo on Mar 12, 2021, 01:07

@feickert I refer to the issues 19, 18, 17 and 13. I should not take more time.
As soon as I've done, I let you know.

**Comment by kratsg** [comment]
In GitLab by @gstark on Mar 31, 2021, 06:14

@llongo -- any updates?

**Comment by kratsg** [comment]
In GitLab by @lheinric on Apr 14, 2021, 13:11

I'm also interested in getting this in, as I have another PR I would like to contribute afterwards. Any way we can help @llongo?

**Comment by kratsg** [comment]
In GitLab by @gstark on May 25, 2021, 07:42

@llongo any progress?

**Comment by kratsg** [comment]
In GitLab by @llongo on May 26, 2021, 15:22

Hi @gstark ,
I've discovered there is an other bug for the recycling of the histograms when an uncertainty with a different nominal tree is used. I'll try to finish it by sunday, if I'll not manage I'll create the 2 tags in such a way we could proceed. I'll try later to do a second merge request with the needed fixes.
As I said, this merge request is ready and I'm really sorry for the very long delay.
Cheers, 
Luigi

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 6, 2021, 22:26

mentioned in merge request !75

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 7, 2021, 16:20

mentioned in merge request !76

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 9, 2021, 20:43

added 27 commits

<ul><li>3d3a34ac...e9b030f0 - 6 commits from branch <code>HistFitter:master</code></li><li>8405bac2 - Update README to reflect v0.66.0</li><li>ab11d9b0 - Add Python requirements.txt files</li><li>35e6950b - Add pre-commit hooks</li><li>733c4317 - Ignore VSCode project settings</li><li>df5f912d - Apply 2to3 and pyupgrade to python/ dir</li><li>68a015eb - Apply 2to3 and pyupgrade to scripts/ dir</li><li>a6c5ea2f - Apply 2to3 and pyupgrade to test/ dir</li><li>f2168b72 - Apply 2to3 and pyupgrade to analysis/ dir</li><li>94351ebb - Apply 2to3 and pyupgrade to doxypy</li><li>1967662f - Remove vendored Python 2.7 argparse</li><li>1557836d - Migrate Dockerfile to Python 3.6+ and ROOT v6.20</li><li>a47f27d2 - Update CI to use Kaniko workflow and add linting</li><li>e0b9c7dd - Reduce abuse of PYTHONPATH</li><li>03b00bca - Set minimum compat ROOT in setup.sh</li><li>68c76caa - Deprecate ROOT.Double() for ctypes.c_double()</li><li>3f968b61 - Use sorted with key option for broader ROOT support</li><li>ff6b1c33 - Add test of analysis/tutorial/MyUserAnalysis.py</li><li>9557ec3b - Fix sorting of InputTrees with filename lambda</li><li>d356a5ea - Fix SysTable and UpperLimitTable namespace error</li><li>75e366e6 - Fix writing of pickle binary</li><li>ea886d72 - Give ROOT ownership of configMgr.chains[chainID]</li></ul>

[Compare with previous version](/HistFitter/HistFitter/-/merge_requests/70/diffs?diff_id=942657&start_sha=3d3a34acf0f8c4469f987108ac3aa53021914330)

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 9, 2021, 20:48

@llongo I've rebased this MR on top of `v0.66.0` and on MR !76, so it should be good to go at this point. :thumbsup: 

Please let me know if there are any additional checks that you'd like to see, but if not I think this is ready.

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 9, 2021, 20:59

changed the description

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 9, 2021, 20:59

changed title from **feat: Migrate codebase to Python 3.6+** to **feat: Migrate codebase to Python 3.6+{+ (!70)+}**

**Comment by kratsg** [comment]
In GitLab by @llongo on Jun 13, 2021, 15:38

Hi @feickert , thanks a lot for all the work. I've added a couple of other lines and modify the ci and the readme.
Tomorrow I'll inform the community about the latest tags.
Thanks,
Luigi

**Comment by kratsg** [comment]
In GitLab by @feickert on Jun 13, 2021, 15:44

Sounds great! Thanks for your reviews and shepherding this MR through @llongo.
